cli/azd/internal/telemetry/uploader.go

```diff
@@ -0,0 +1,193 @@
+	var item *StoredItem
+	err := retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		peekItem, err := u.telemetryQueue.Peek()
+
+		if err != nil {
+			return retry.RetryableError(err)
+		}
+
+		item = peekItem
+		return nil
+	})
+
+	if err != nil && ctx.Err() != nil {
+		// Attempt fallback - remove and retry Peek
+		err = u.reliableRemove(ctx, item)
+
+		if err != nil {
+			item, err = u.telemetryQueue.Peek()
+		}
+	}
+
+	return item, err
+}
+
+func (u *TelemetryUploader) reliableRemove(ctx context.Context, item *StoredItem) error {
+	return retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		return retry.RetryableError(u.telemetryQueue.Remove(item))
+	})
+}
+
+func (u *TelemetryUploader) reliableEnqueueWithDelay(ctx context.Context, itemName string, payload []byte, delayDuration time.Duration, attempts int) {
+	err := retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		return retry.RetryableError(u.telemetryQueue.EnqueueWithDelay(payload, delayDuration, attempts))
+	})
+
+	if err != nil {
+		log.Printf("unable to requeue item %v after %d attempts", itemName, attempts)
+	}
+}
+
+func (u *TelemetryUploader) transmit(ctx context.Context, item *StoredItem) {
+	payload := item.Message()
+	var telemetryItems appinsightsexporter.TelemetryItems
+	if u.isDebugMode {
+		// When in debug mode, we deserialize to get better error messages
+		telemetryItems.Deserialize(payload)
+	}
+	result, err := u.transmitter.Transmit(payload, telemetryItems)
+	if err == nil && result != nil && result.IsSuccess() {
+		return
+	}
+
+	attempts := item.RetryCount() + 1
+
+	if err != nil || result == nil {
+		if attempts > maxRetryCount {
+			log.Printf("failed to send %v after %d attempts, err: %v\n", item.fileName, maxRetryCount, err)
+			return
+		}
+
+		u.reliableEnqueueWithDelay(ctx, item.fileName, payload, defaultTransmitRetryDelay, attempts)
+	} else if result.CanRetry() {
+		if attempts > maxRetryCount {
+			log.Printf("failed to send %v after %d attempts, statusCode: %d\n", item.fileName, maxRetryCount, result.StatusCode)
+			return
+		}
+
+		if result.IsThrottled() {
+			var throttleDuration time.Duration
+			if result.RetryAfter != nil {
+				throttleDuration = u.clock.Until(*result.RetryAfter)
+			} else {
+				throttleDuration = defaultThrottleDuration
+			}
+
+			log.Printf("Upload is being throttled. Resuming upload in %v.", throttleDuration)
+			time.Sleep(throttleDuration)
+		}
+
+		retryDelay := u.calculateRetryDelay(result.RetryAfter, attempts)
+
+		if result.IsPartialSuccess() {
+			var telemetryItems appinsightsexporter.TelemetryItems
+			telemetryItems.Deserialize(payload)
+			newPayload, _ := result.GetRetryItems(payload, telemetryItems)
```

Comment:
Just checking - if `GetRetryItems` were to return an error, the first value would still be something reasonable to pass to `reliableEnqueueWithDelay`, correct?

Reply:
The second value returned here is not an `error`. It is the unmarshaled struct.

The one error case is if the server sent an invalid reply containing invalid error indices. In such a case, `GetRetryItems` would return the original payload unfiltered.